src/main/kotlin/org/rust/lang/core/psi/util/RustPsiNavigationExtensions.kt

```diff
@@ -0,0 +1,159 @@
+package org.rust.lang.core.psi.util
+
+import com.intellij.psi.PsiComment
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiWhiteSpace
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.psi.util.PsiUtilCore
+import org.rust.lang.core.psi.RustBlockElement
+import org.rust.lang.core.psi.RustExprElement
+import org.rust.lang.core.psi.RustStmtElement
+
+inline fun <reified T : PsiElement> PsiElement.parentOfType(strict: Boolean = true, minStartOffset: Int = -1): T? =
+    PsiTreeUtil.getParentOfType(this, T::class.java, strict, minStartOffset)
+
+inline fun <reified T : PsiElement> PsiElement.childOfType(strict: Boolean = true): T? =
+    PsiTreeUtil.findChildOfType(this, T::class.java, strict)
+
+inline fun <reified T : PsiElement> PsiElement.descendentsOfType(): Collection<T> =
+    PsiTreeUtil.findChildrenOfType(this, T::class.java)
+
+
+/**
+ * Finds first sibling that is neither comment, nor whitespace before given element.
+ */
+fun PsiElement?.getPrevNonCommentSibling(): PsiElement? =
+    PsiTreeUtil.skipSiblingsBackward(this, PsiWhiteSpace::class.java, PsiComment::class.java)
+
+/**
+ * Finds first sibling that is neither comment, nor whitespace after given element.
+ */
+fun PsiElement?.getNextNonCommentSibling(): PsiElement? =
+    PsiTreeUtil.skipSiblingsForward(this, PsiWhiteSpace::class.java, PsiComment::class.java)
+
+
+/**
+ * Finds top-most [RustExprElement] within selected range.
+ * Supports such cases: `ident&lt;if&gt;ier` (or keyword if applicable),
+ * `&lt;&nbsp;&nbsp;&nbsp;expression&nbsp;&nbsp;&nbsp;&gt;`.
+ */
+fun findExpressionInRange(file: PsiFile, startOffset: Int, endOffset: Int): RustExprElement? {
+    val (element1, element2) = file.getElementRange(startOffset, endOffset) ?: return null
+
+    // Get common expression parent.
+    var parent = PsiTreeUtil.findCommonParent(element1, element2) ?: return null
+    if (parent !is RustExprElement) {
+        parent = parent.parentOfType<RustExprElement>() ?: return null
+    }
+
+    // If our parent's deepest first child is element1 and deepest last - element 2,
+    // then is is completely within selection, so this is our sought expression.
+    if (element1 == PsiTreeUtil.getDeepestFirst(parent) && element2 == PsiTreeUtil.getDeepestLast(element2)) {
+        return parent
+    }
+
+    return null
+}
+
+// FIXME: Maybe items are ok?
+// FIXME: What about macros?
+/**
+ * Finds statements (mostly [RustStmtElement]s, [PsiComment]s and [RustExprElement]
+ * as return expression, doesn't allow attrs and items) within selected range.
+ */
+fun findStatementsInRange(file: PsiFile, startOffset: Int, endOffset: Int): Array<out PsiElement> {
```

Comment:
API requirement of `SurroundDescriptor#getElementsToSurround()`. This array must be created anyway